### PR TITLE
Reliable uploads -  UPLOADS VIEW

### PR DESCRIPTION
**Pull request by davivel**
@owncloud/android-developers 

Time has come.

From my POV, this is finally ready to be tested by QA (cc @jesmrec).

Many hands contributed here, starting with the awesome effort of @LukeOwncloud. Sounds wise summoning many eyes to review it. So, any comment about the code will be welcome.

Though many things have been rewritten and considerably cleaned-up, sure you will still find many points that could (and should) be refactored. I can't promise that all of them may be attended in this PR, since we really, really, need to release this with next version 1.9.2, that should be very near in time to the release of OC 9. Evolution of testing is the leading criteria right now. 

This means: if we think we can refactor something in a safe manner without introducing an important risk for QA, we will do. If bugs appear and the code to fix is also subject of comments in the code review, we will apply the comments. 

But if change requests are big enough to require QA starts the testing all over again, we will have to wait for a better chance.

This one should not be too far in time. This PR lets aside an important topic that was originally addressed by @LukeOwncloud , the improvement of automatic retries. Right now instant uploads are retried when Wifi connection is established, in a similar way to the operation of the currently released version 1.9.1 [1]. But other convenient improvements will be addressed specifically in the short term [2], not in this PR. And that should be a good moment to improve the code of uploads further.

Come on Barbie, let's go party!

[1] You may notice some light improvements, but that's a side effect of the many changes done; was easier let it that way that adding artificial limits. Have fun finding them :)

[2] If you don't want to wait and prefer to implement them, go for it. Just mention me to let me know, and I'll try to help.

----

BUGS FOUND AND ENHANCEMENT:

- [x] Failed uploads can not be retried in another account #1497
- [x] Failed uploads are not retried by the button #1498
- [x] Deleting an account does not clear the current uploads #1499 
- [x] Killing the app leaves uploads in an inconsistent state #1500
- [x] Wording issues in uploads view #1501
- [x] Manual retries are "delayed #1502
- [x] Uploads in SAML requests authentication twice #1503
- [x] Current upload should be in the top of the uploads view #1511
- [x] Only current upload in progress naming as "Uploading" #1520
- [x] Retry failed uploads start uploading by the bottom #1521
- [x] Credentials requested on each failed upload #1522 
- [x] Uploaded files are possible to uploaded it again #1523
- [x] Title on top bar in upload view is shown sometimes owncloud instead of Uploads #1524 
- [x] Failed Uploads set as wrong status "Waiting for wifi connectivity"  #1525 
- [x] Show account info also while the file is uploading or waiting to upload #1526 
- [x] Title on top bar in upload view is shown the name of the failed upload #1530 
- [x] Some Instant uploads naming as "Unknown error" #1531
- [x] Uploads with status "Waiting wifi connectivity" are not retried #1532
- [x] Clear all uplodas does not cancel all the current uploads #1533 
- [x] Current upload can't be cancelled (Android 4.1.2) #1535
- [x] Progress bar not updated in uploads view when the upload is retried manually (Android 4.1.2) #1536
- [x] Upload the same file to different folder fails #1549 
- [x] Deleting a file while uploading results in "Connection error" message instead of "file deleted locally" #1528 
- [x] Background color should be white everywhere #1426
- [x] Thumbnails not shown in `UPLOADED` section #1517 
- [x] If the same file is uploaded twice, both uploads are removed from list #1553 
- [x] Yellow arrow when the file has a duplicated name #1554 [OUT OF SCOPE]
- [x] If wifi is lost, only one upload is move to failed #1569 
- [x] Username in drawer not refreshed when account is removed from uploads view #1572
- [x] When a lot of files is sent to uploads view, progress bar is not correctly managed #1573 
- [x] Upload path does not appear in "Uploaded" #1576

### CODE REVIEW

**Comment by tobiasKaminsky** [comment]
This in now in Beta :)

Some thoughts about it:
- no thumbnails for images? (can I just add this? or should I create a successor branch of it?)
- "current" and "uploaded" should show total number of files: current (75)
- "uploading..." has no percentage whereas the notification has

Looking forward to test the view if uploads are failing ;)

**Comment by LukeOwncloud** [comment]
Should we report errors here? Or create a github issue? 

Issue: Kill owncloud app via swipe out while upload in progress leaves app in inconsistent state. Notification gets stuck. Upload view may show upload both in "current" and "uploaded" sections.

**Comment by davivel** [comment]
> This in now in Beta :)

Awesome!

> no thumbnails for images? (can I just add this? or should I create a successor branch of it?)

Mmm, that should be working; maybe it's not refreshing correctly, but I remember the code is there.

> 0B in "uploaded" (if file is moved to owncloud)

That's a bug, @jesmrec also found it. We will record the bugs found internally in Github to reduce overlapping.

> Should we report errors here? Or create a github issue?

A Github issue per bug, please. Then we link it here, in the first comment whether possible.

> "current" and "uploaded" should show total number of files: current (75)
> "uploading..." has no percentage whereas the notification has

Hey, those are feature requests :P ;) . We can consider, since they're pretty small, but need to be careful with extending the scope. 

Rest of your comments could be bugs, let's have a look.

**Comment by jesmrec** [comment]
You can also track the QA process:

https://github.com/owncloud/QA/issues/56

**Comment by davivel** [comment]
> You can also track the QA process:
> owncloud/QA#56

Definitely, yes!

**Comment by davivel** [comment]
> can I just add this?

FYI, we'll be very glad to receive any help in bug fixing. Just assign to yourself the corresponding issue in Github before working in it so that we don't overlap. If the fix is tiny it's OK to push directly in the same branch (please, use ```git rebase``` if find new commits in remote in the moment of pushing). If the changes are big, better creating a successor branch and create PR on ```reliable_uploads_actions_uploads_view```.

**Comment by enoch85** [comment]
Finally! :D Will this be in 1.9.3?

**Comment by AndyScherzinger** [comment]
See @davivel's original comment, if everything goes according to plan than it'll be released with 1.9.2 :grin:

**Comment by enoch85** [comment]
Amazing news! :100:

**Comment by AndyScherzinger** [comment]
Looking at the number of bugs found that need fixing it is still hard to say if we can make it happen

**Comment by tobiasKaminsky** [comment]
@davivel this bot is annoying :/
@DeepDiver1975 @masensio is not known to the bot?

**Comment by AndyScherzinger** [comment]
@jancborchardt in oarticular and to @masensio and @davivel in general, as mentioned in my comments and agreed on by Jan, I made some alignment changes to the uploader layout (not commited yet) for you to review.

Right now the layout looks like this (actual branch, *without* my changes)
device-2016-03-14-104935

With my minor changes the layout looks like this (basically aligning icons and text with the action bar):
device-2016-03-14-112203

As soon as I get positive feedback from you guys I can commit/push the changes.

**Comment by davivel** [comment]
> Looking at the number of bugs found that need fixing it is still hard to say if we can make it happen

@AndyScherzinger , 1.9.2 will not be released without this PR. It's a must. :-)

**Comment by AndyScherzinger** [comment]
Uuuh, nice. Okay. So @enoch85 it is def. 1.9.2 :+1: 

@davivel any comments on the screenshots? :)

**Comment by davivel** [comment]
@AndyScherzinger , push those alignment changes in the branch if you wish. Thanks

**Comment by AndyScherzinger** [comment]
@davivel DONE - pushed the changes

**Comment by DeepDiver1975** [comment]
Why are there so many commits in this pr? Maybe a rebase will clean it up?

**Comment by enoch85** [comment]
Great! @AndyScherzinger :D Any release date @davivel ?

**Comment by davivel** [comment]
@DeepDiver1975 a rebase at this point is madness. We tried.

**Comment by davivel** [comment]
@AndyScherzinger , no problem about the background. 

I'm more concerned about the last commit with the merge. You can prevent those extra commits using ```git rebase origin/reliable_uploads_actions_view_uploads``` instead of ```git merge origin/reliable_uploads_actions_view_uploads```. Come on, you can do it! :)

**Comment by AndyScherzinger** [comment]
I didn't intend to do a merge (again... :disappointed:) I am using tortoiseGit and for when I tried to commit is was behind the latest revision and did a pull on the branch which should not lead to a merge imho and then did a push. Totally confused as to why this led to a merge of the branch with itself (or I still don't get it which seems likely to be the case...)

**Comment by davivel** [comment]
In git, pull = (fetch + merge). That's the point. :-)

**Comment by AndyScherzinger** [comment]
Thanks for clearing that up, didn't realize that since I always used pull and it to my knowledge it never led to a merge :S Will try my very best getting to cope with rebase :)

**Comment by davivel** [comment]
@jesmrec , at this point all the reported bugs are fixed. Let's test them and finish this :-)

**Comment by jesmrec** [comment]
Perfect @davivel, great job! :) the nice feature is closer to be finished.

**Comment by AndyScherzinger** [comment]
@davivel your dev speed at this very moment regarfing the Android app...Wow, your the bomb! :+1: :+1: :+1: This is awesome! :tada:

**Comment by enoch85** [comment]
:beers: Remind me to buy you a beer next time we see eachother @davivel :)

**Comment by davivel** [comment]
@enoch85, @AndyScherzinger , @all, thanks, we can feel your love :-) 

Not working specially harder than usual, just very, very, very focused in this feature, and tracking the bug-fixing progress more openly than usual.

**Comment by enoch85** [comment]
> and tracking the bug-fixing progress more openly than usual.

@davivel Would be great if you were able to keep it that way :)

**Comment by davivel** [comment]
We'll try :)

**Comment by strugee** [comment]
@AndyScherzinger you should check out `branch.autoSetupRebase`, `pull.rebase`, and `branch.<name>.rebase` in the [git-config(1)][1] manpage :)

 [1]: https://www.kernel.org/pub/software/scm/git/docs/git-config.html

File: src/com/owncloud/android/files/services/FileUploader.java [hunk 1]
```diff
@@ -659,7 +660,7 @@ private void cancel(String accountName, String remotePath) {
             UploadFileOperation upload = removeResult.first;
             if (upload == null &&
                     mCurrentUpload != null && mCurrentAccount != null &&
-                    mCurrentUpload.getRemotePath().startsWith(remotePath) &&
+                    remotePath.contains(mCurrentUpload.getFileName()) &&
```
Comment (davivel) [inline src/com/owncloud/android/files/services/FileUploader.java]: This is not a good replacement.

**Comment by jancborchardt** [comment]
@AndyScherzinger I would remove the »delete« icon/function from the sucessfully uploaded files. It’s not necessary to have and on top of this it _will_ confuse people thinking it deletes the file.

What we however need is a »Retry« button there for failed uploads. Do you have a screenshot of that case?

**Comment by AndyScherzinger** [comment]
@jancborchardt I don't unfortunately. @davivel can you provide Jan with a screenshot and reply to Jan regarding the delete issue? :smiley:

**Comment by AndyScherzinger** [comment]
@DeepDiver1975 it seems the CLA Bot has a problem resolving @masensio's commits or her mail address respectively. This is spamming the PR threads :cry:

**Comment by davivel** [comment]
> Do you have a screenshot of that case?

Sorry, @jancborchardt , I overlooked your comment :S 

See here a couple of current screenshots.

upload_view_20160323

uploads_menu

> I would remove the »delete« icon/function from the successfully uploaded files.

We can do that. That option it's not available in iOS, either. And that reminds me that we should change the way the maximum of uploads is handled, it's being filtered in the view, while should be a maximum in storage. I'll add an issue for both.

> What we however need is a »Retry« button there for failed uploads. 

In a moment we had the refresh button with the trash button together.  Seemed overloaded, and misaligned with the other section, where items only have a button. In iOS a tap on the items triggers the 'fix & retry', so we took the same approach. 

Recently I proposed to add (tap to retry) in the section header so that it's clearer, but I'm not sure it was a good idea, it's ugly as hell. I'm open to review this part of the UI, but we should decide soon, testing is again in "hardcode mode" on this PR, and we need to close it ASAP.

**Comment by davivel** [comment]
@jancborchardt , sorry, I'm trying to find a commit with both buttons 'refresh' and 'trash' in a single item to visually compare, but I can't find it.

**Comment by enoch85** [comment]
@jancborchardt @davivel 
> What we however need is a »Retry« button there for failed uploads. 

"Retry all" is what you'd want anyway. Not tapping each failed download. Or am I misunderstanding here?

File: src/com/owncloud/android/datamodel/UploadsStorageManager.java [hunk 2]
```diff
@@ -515,14 +516,16 @@ public void updateDatabaseUploadResult(RemoteOperationResult uploadResult, Uploa
                         upload.getOCUploadId(),
                         UploadStatus.UPLOAD_SUCCEEDED,
                         UploadResult.UPLOADED,
-                        upload.getRemotePath()
+                        upload.getRemotePath(),
+                        upload.getStoragePath()
```
Comment (davivel) [inline src/com/owncloud/android/datamodel/UploadsStorageManager.java]: Will not be so easy; this leads to a crash (later) if behaviour for instant uploads is set to "forget"

**Comment by tobiasKaminsky** [comment]
@enoch85 There is "retry failed", which should retry all failed.

**Comment by enoch85** [comment]
@tobiasKaminsky Yes I saw that, but what I mean is that the action "retry only this file" should not replace "retry all".

**Comment by davivel** [comment]
>  what I mean is that the action "retry only this file" should not replace "retry all".

True. No problem, both live happy and together, one in the floating menu, the other tapping on the target element.

**Comment by enoch85** [comment]
@davivel Great! Any ETA on this branch?

**Comment by davivel** [comment]
> I would remove the »delete« icon/function from the successfully uploaded files.

@jancborchardt , DONE now.

> that reminds me that we should change the way the maximum of uploads is handled, it's being filtered in the view, while should be a maximum in storage.

DONE now

> Any ETA on this branch?

@enoch85 , tomorrow in master, in @jesmrec agrees :) 

Then, for 1.9.2, we need to fix #1514 , and enter in "release mode".  Next week if we are lucky with the regression tests.

**Comment by jesmrec** [comment]
Approved. Great job of all the android team & community!!!! Congrats

**Comment by davivel** [comment]
:monkey: :dancer:

**Comment by rperezb** [comment]
:trophy:

**Comment by davivel** [comment]
Last merge from master to fix conflict was totally trivial.

:+1: to all the code not written by me.

@malkomich , please, review code written by me.

**Comment by davivel** [comment]
@DeepDiver1975 , please, we need to remove the ```Required``` status of ```cla-bot-android``` until we can fix or prevent the current error.

**Comment by malkomich** [comment]
:+1:

**Comment by davivel** [comment]
:tophat:

**Comment by AndyScherzinger** [comment]
Aweseome! Kudos to everyone involved in this Mega-PR! :tada: :clap:

**Pull request merged**